core/src/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,700 @@
+    num_dropped_gossip: usize,
+    num_dropped_tpu: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct LatestUnprocessedVotes {
+    latest_votes_per_pubkey: RwLock<HashMap<Pubkey, RwLock<RefCell<DeserializedVotePacket>>>>,
+}
+
+unsafe impl Send for LatestUnprocessedVotes {}
+unsafe impl Sync for LatestUnprocessedVotes {}
+
+impl LatestUnprocessedVotes {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Expensive
+    pub fn len(&self) -> usize {
+        self.latest_votes_per_pubkey
+            .read()
+            .unwrap()
+            .values()
+            .filter(|lock| {
+                lock.read()
+                    .ok()
+                    .map(|vp| !vp.borrow().is_processed())
+                    .unwrap_or(false)
+            })
+            .count()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn insert_batch(
+        &self,
+        votes: impl Iterator<Item = DeserializedVotePacket>,
+    ) -> VoteBatchInsertionMetrics {
+        let mut num_dropped_gossip = 0;
+        let mut num_dropped_tpu = 0;
+
+        for vote in votes {
+            if let Some(vote) = self.update_latest_vote(vote) {
+                match vote.vote_source {
+                    VoteSource::Gossip => num_dropped_gossip += 1,
+                    VoteSource::Tpu => num_dropped_tpu += 1,
+                }
+            }
+        }
+
+        VoteBatchInsertionMetrics {
+            num_dropped_gossip,
+            num_dropped_tpu,
+        }
+    }
+
+    /// If this vote causes an unprocessed vote to be removed, returns Some(old_vote)
+    /// If there is a newer vote processed / waiting to be processed returns Some(vote)
+    /// Otherwise returns None
+    pub fn update_latest_vote(
+        &self,
+        vote: DeserializedVotePacket,
+    ) -> Option<DeserializedVotePacket> {
+        let pubkey = vote.pubkey();
+        let slot = vote.slot();
+        if let Some(latest_vote) = self.latest_votes_per_pubkey.read().unwrap().get(&pubkey) {
+            let latest_slot = latest_vote
+                .read()
+                .unwrap()
+                .try_borrow()
+                .ok()
+                .map(|vote| vote.slot())
+                .unwrap_or(0);
+            if slot > latest_slot {
+                let latest_vote = latest_vote.write().unwrap();
+                // At this point no one should have a borrow to this refcell as all borrows are
+                // hidden behind read()
+                if let Ok(mut latest_vote) = latest_vote.try_borrow_mut() {
+                    let latest_slot = latest_vote.slot();
+                    if slot > latest_slot {
+                        let old_vote = std::mem::replace(latest_vote.deref_mut(), vote);
+                        if old_vote.is_processed() {
+                            return None;
+                        } else {
+                            return Some(old_vote);
+                        }
```

Comment:
once we replace `clear()` with `take_vote()`, this can just be replaced with `old_vote.take_vote()`

Reply:
We want to return the whole packet though, `take_vote` will only return the immutable section.